src/Shared/WeakStringCache.cs

```diff
@@ -0,0 +1,355 @@
+        /// adds the resulting string to the cache, and returns it setting cacheHit to false.
+        /// </summary>
+        /// <param name="internable">The internable describing the string we're looking for.</param>
+        /// <returns>A string matching the given internable.</returns>
+        public string GetOrCreateEntry<T>(T internable, out bool cacheHit) where T : IInternable
+        {
+            int hashCode = GetInternableHashCode(internable);
+
+            StringBucket bucket;
+            string result;
+            lock (_stringsByHashCode)
+            {
+                if (!_stringsByHashCode.TryGetValue(hashCode, out bucket))
+                {
+                    bucket = new StringBucket();
+                }
+                result = bucket.GetString(internable);
+                if (result != null)
+                {
+                    cacheHit = true;
+                    return result;
+                }
+            }
+
+            // We don't have the string in the dictionary - create it.
+            result = internable.ExpensiveConvertToString();
+
+            lock (_stringsByHashCode)
+            {
+                bucket.AddString(result);
+
+                // Prevent the dictionary from growing forever with buckets whose underlying GC handles
+                // don't reference any live strings anymore.
+                if (_stringsByHashCode.Count >= _capacity)
+                {
+                    // Get rid of unused buckets.
+                    Scavenge();
+                    // And do this again if the number of buckets reaches double the current after-scavenge number.
+                    _capacity = _stringsByHashCode.Count * 2;
+                }
+                _stringsByHashCode[hashCode] = bucket;
+            }
+            cacheHit = false;
+            return result;
+        }
+
+        /// <summary>
+        /// Implements the simple yet very decently performing djb2 hash function.
+        /// </summary>
+        /// <param name="internable">The internable to compute the hash code for.</param>
+        /// <returns>The 32-bit hash code.</returns>
+        internal static int GetInternableHashCode<T>(T internable) where T : IInternable
+        {
+            int hashCode = 5381;
+            for (int i = 0; i < internable.Length; i++)
+            {
+                unchecked
+                {
+                    hashCode = hashCode * 33 + internable[i];
+                }
+            }
+            return hashCode;
+        }
+
+        /// <summary>
+        /// Iterates over the cache and removes unused buckets, i.e. buckets that don't reference live strings.
+        /// This is expensive so try to call such that the cost is amortized to O(1) per GetOrCreateEntry() invocation.
+        /// Assumes exclusive access to the dictionary, i.e. the lock is taken.
+        /// </summary>
+        public void Scavenge()
```

Comment:
Should this be public? How can an external caller guarantee the lock on `_stringsByHashCode` is held?

Reply:
Great point, thanks. It's used only by tests so far but that's definitely a bug. Fixed by splitting to a public `Scavenge` and non-public `ScavengeUnderLock`.